File: src/csharp/pulse/cdm/scalars/SEScalarPower.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

public class PowerUnit
{
  private PowerUnit(string v) { Value = v; }
  protected string Value { get; }

  public new string ToString() { return Value; }

  public static PowerUnit W = new PowerUnit("W");
  public static PowerUnit kcal_Per_s = new PowerUnit("kcal/s");
  public static PowerUnit kcal_Per_hr = new PowerUnit("kcal/hr");
  public static PowerUnit kcal_Per_day = new PowerUnit("kcal/day");
  public static PowerUnit J_Per_s = new PowerUnit("J/s");
  public static PowerUnit BTU_Per_hr = new PowerUnit("BTU/hr");

  public static PowerUnit FromString(string u)
  {
    if (u == PowerUnit.W.ToString())
      return PowerUnit.W;
    if (u == PowerUnit.kcal_Per_s.ToString())
      return PowerUnit.kcal_Per_s;
    if (u == PowerUnit.kcal_Per_hr.ToString())
      return PowerUnit.kcal_Per_hr;
    if (u == PowerUnit.kcal_Per_day.ToString())
      return PowerUnit.kcal_Per_day;
    if (u == PowerUnit.J_Per_s.ToString())
      return PowerUnit.J_Per_s;
    if (u == PowerUnit.BTU_Per_hr.ToString())
      return PowerUnit.BTU_Per_hr;

    throw new System.ArgumentException(u + " is not a know power unit");
  }
}

public class SEScalarPower : SEScalar
{
  protected PowerUnit unit;

  public SEScalarPower() : base()
  {

  }

  public SEScalarPower(SEScalarPower from)
  {
    Set(from);
  }

  public SEScalarPower(double value, PowerUnit unit)
  {
    SetValue(value, unit);
  }

  public void Set(SEScalarPower from)
  {
    if (from == null)
    {
      Invalidate();
      return;
    }
    SetValue(from.value, from.unit);
  }

  public new void SetValue(double value)
  {
    throw new System.NotSupportedException("You must provide a unit");
  }

  public void SetValue(double value, PowerUnit unit)
  {
    this.value = value;
    this.unit = unit;
  }

  public double GetValue(PowerUnit unit)
  {
    throw new System.NotImplementedException();
  }

  public PowerUnit GetUnit()
  {
    return unit;
  }

  public new string ToString()
  {
    return this.value + "(" + this.unit.ToString() + ")";
  }
}

File: src/cpp/cdm/properties/SEScalarPower.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarPower.h"

const PowerUnit PowerUnit::W("W");
const PowerUnit PowerUnit::kcal_Per_s("kcal/s");
const PowerUnit PowerUnit::kcal_Per_hr("kcal/hr");
const PowerUnit PowerUnit::kcal_Per_day("kcal/day");
const PowerUnit PowerUnit::J_Per_s("J/s");
const PowerUnit PowerUnit::BTU_Per_hr("BTU/hr");

bool PowerUnit::IsValidUnit(const std::string& unit)
{
  if (W.GetString().compare(unit) == 0)
    return true;
  if (kcal_Per_s.GetString().compare(unit) == 0)
    return true;
  if (kcal_Per_hr.GetString().compare(unit) == 0)
    return true;
  if (kcal_Per_day.GetString().compare(unit) == 0)
    return true;
  if (J_Per_s.GetString().compare(unit) == 0)
    return true;
  if (BTU_Per_hr.GetString().compare(unit) == 0)
    return true;
  return false;
}

const PowerUnit& PowerUnit::GetCompoundUnit(const std::string& unit)
{
  if (W.GetString().compare(unit) == 0)
    return W;
  if (kcal_Per_s.GetString().compare(unit) == 0)
    return kcal_Per_s;
  if (kcal_Per_hr.GetString().compare(unit) == 0)
    return kcal_Per_hr;
  if (kcal_Per_day.GetString().compare(unit) == 0)
    return kcal_Per_day;
  if (J_Per_s.GetString().compare(unit) == 0)
    return J_Per_s;
  if (BTU_Per_hr.GetString().compare(unit) == 0)
    return BTU_Per_hr;
  throw CommonDataModelException(unit + " is not a valid Power unit");
}

template class SEScalarQuantity<PowerUnit>;
